ipld/get_namespaced_shares.go

```diff
@@ -0,0 +1,226 @@
+// and we don't know in advance how many jobs we will have to wait for.
+type wrappedWaitGroup struct {
+	wg      sync.WaitGroup
+	jobs    chan *job
+	counter int64
+}
+
+func (w *wrappedWaitGroup) Add(count int64) {
+	w.wg.Add(int(count))
+	atomic.AddInt64(&w.counter, count)
+}
+
+func (w *wrappedWaitGroup) Done() {
+	w.wg.Done()
+	numRemaining := atomic.AddInt64(&w.counter, -1)
+
+	// Close channel if this job was the last one
+	if numRemaining == 0 {
+		close(w.jobs)
+	}
+}
+
+func (w *wrappedWaitGroup) Wait() {
+	w.wg.Wait()
+}
+
+type fetchedBounds struct {
+	lowest  int64
+	highest int64
+}
+
+func (b *fetchedBounds) Update(index int64) {
+	lowest := atomic.LoadInt64(&b.lowest)
+	for index < lowest && !atomic.CompareAndSwapInt64(&b.lowest, lowest, index) {
+		lowest = atomic.LoadInt64(&b.lowest)
+	}
+	// we always run both checks because element can be both the lower and higher bound
+	// for example, if there is only one share in the namespace
+	highest := atomic.LoadInt64(&b.highest)
+	for index > highest && !atomic.CompareAndSwapInt64(&b.highest, highest, index) {
+		highest = atomic.LoadInt64(&b.highest)
+	}
+}
+
+// getLeavesByNamespace returns all the leaves from the given root with the given namespace.ID.
+// If no shares are found, it returns both data and error as nil.
+// A non-nil error means that only partial data is returned, because at least one share retrieval failed
+// The following implementation is based on `GetShares`.
+func getLeavesByNamespace(
+	ctx context.Context,
+	bGetter blockservice.BlockGetter,
+	root cid.Cid,
+	nID namespace.ID,
+	maxShares int,
+) ([]ipld.Node, error) {
+	ctx, span := tracer.Start(ctx, "get-leaves-by-namespace")
+	defer span.End()
+
+	span.SetAttributes(
+		attribute.String("namespace", nID.String()),
+		attribute.String("root", root.String()),
+	)
+
+	err := SanityCheckNID(nID)
+	if err != nil {
+		return nil, err
+	}
+
+	// we don't know where in the tree the leaves in the namespace are,
+	// so we keep track of the bounds to return the correct slice
+	// maxShares acts as a sentinel to know if we find any leaves
+	bounds := fetchedBounds{int64(maxShares), 0}
+
+	// buffer the jobs to avoid blocking, we only need as many
+	// queued as the number of shares in the second-to-last layer
+	jobs := make(chan *job, (maxShares+1)/2)
+	jobs <- &job{id: root}
+
+	var wg wrappedWaitGroup
+	wg.jobs = jobs
+	wg.Add(1)
+
+	var singleErr sync.Once
+	var retrievalErr error
+
+	// we overallocate space for leaves since we do not know how many we will find
+	// on the level above, the length of the Row is passed in as maxShares
+	leaves := make([]ipld.Node, maxShares)
+
+	// stillWorking will be set to false when the waitgroup counter reaches 0, closing the jobs channel
+	for {
+		select {
+		case j, ok := <-jobs:
+			if !ok {
+				// we didn't find any leaves, return nil
+				if bounds.lowest == int64(maxShares) {
+					return nil, retrievalErr
```

Comment:
We should document the case where both nils are returned here. When there is no namespace under the given root

Reply:
@Wondertan does the new comment doc this as intended?